File: app/src/main/java/com/diamond/appcliente/actividades/UsuarioActivity.java
package com.diamond.appcliente.actividades;

import android.content.Intent;
import android.net.Uri;
import android.os.Bundle;
import android.provider.MediaStore;
import android.util.Log;
import android.view.MenuItem;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Button;
import android.widget.EditText;
import android.widget.ImageView;
import android.widget.PopupWindow;
import android.widget.TextView;
import android.widget.Toast;
import androidx.appcompat.app.AppCompatActivity;
import com.auth0.android.jwt.JWT;
import com.bumptech.glide.Glide;
import com.diamond.appcliente.R;
import com.diamond.appcliente.dto.usuario.UsuarioDto;
import com.diamond.appcliente.util.PreferenciasHelper;
import com.diamond.appcliente.viewmodel.UsuarioViewModel;
import com.google.android.material.bottomnavigation.BottomNavigationView;

public class UsuarioActivity extends AppCompatActivity {
    private Button btnLogoutUser;
    private ImageView imageUsuario;
    private Uri imagenSeleccionadaUri;
    private TextView textCelular;
    private TextView textEmail;
    private TextView textNombre;
    private UsuarioViewModel viewModel;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_usuario);
        imageUsuario = findViewById(R.id.imageUsuario);
        textNombre = findViewById(R.id.textNombre);
        textEmail = findViewById(R.id.textEmail);
        textCelular = findViewById(R.id.textCelular);
        btnLogoutUser = findViewById(R.id.btnLogoutUser);
        viewModel = new UsuarioViewModel();

        String token = new PreferenciasHelper(getApplication()).obtenerToken();
        if (token != null) {
            Log.d("UsuarioActivity", "Token recibido: " + token);
            new JWT(token);
            viewModel.obtenerMiUsuario(this, new UsuarioViewModel.UsuarioCallback() {
                @Override
                public void onSuccess(UsuarioDto usuario) {
                    textNombre.setText(usuario.getNombre() + " " + usuario.getApellido());
                    textEmail.setText(usuario.getEmail());
                    textCelular.setText(usuario.getCelular());
                    Glide.with(UsuarioActivity.this).load(usuario.getUrlUsuario()).placeholder(R.drawable.perfil_default).into(imageUsuario);
                }

                @Override
                public void onError(String mensaje) {
                    Toast.makeText(UsuarioActivity.this, "Error al cargar usuario: " + mensaje, Toast.LENGTH_SHORT).show();
                }
            });
        } else {
            Log.d("UsuarioActivity", "Token no encontrado");
        }

        getWindow().setStatusBarColor(-16777216);

        // Usando clases anónimas en lugar de lambdas
        findViewById(R.id.buttonActualizarUsuario).setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                mostrarPopupActualizar();
            }
        });

        btnLogoutUser.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                cerrarSesion();
            }
        });

        BottomNavigationView bottomNavigationView = findViewById(R.id.bottomNavigationView);
        bottomNavigationView.setSelectedItemId(R.id.nav_perfil);
        bottomNavigationView.setOnItemSelectedListener(new BottomNavigationView.OnItemSelectedListener() {
            @Override
            public boolean onNavigationItemSelected(MenuItem item) {
                int id = item.getItemId();
                if (id == R.id.nav_home) {
                    startActivity(new Intent(UsuarioActivity.this, ClienteHomeActivity.class));
                    overridePendingTransition(0, 0);
                    return true;
                } else if (id == R.id.nav_servicios) {
                    startActivity(new Intent(UsuarioActivity.this, SeccionServiciosActivity.class));
                    overridePendingTransition(0, 0);
                    return true;
                } else if (id == R.id.historial) {
                    startActivity(new Intent(UsuarioActivity.this, ListarReservaActivity.class));
                    overridePendingTransition(0, 0);
                    return true;
                } else if (id != R.id.nav_perfil) {
                    return false;
                } else {
                    startActivity(new Intent(UsuarioActivity.this, UsuarioActivity.class));
                    overridePendingTransition(0, 0);
                    return true;
                }
            }
        });
    }

    private void mostrarPopupActualizar() {
        View popupView = getLayoutInflater().inflate(R.layout.popup_actualizar_usuario, null);
        PopupWindow popupWindow = new PopupWindow(popupView, ViewGroup.LayoutParams.MATCH_PARENT, ViewGroup.LayoutParams.WRAP_CONTENT, true);

        final EditText editTextNombre = popupView.findViewById(R.id.etNombreUsuario);
        final EditText editTextApellido = popupView.findViewById(R.id.etApellidoUsuario);
        final EditText editTextEmail = popupView.findViewById(R.id.etEmailUsuario);
        final EditText editTextCelular = popupView.findViewById(R.id.etCelularUsuario);
        Button btnActualizarImagen = popupView.findViewById(R.id.btnSeleccionarImagenUsuario);
        Button btnActualizarDatos = popupView.findViewById(R.id.btnActualizarUsuario);
        Button btnCancelarActualizarUsuario = popupView.findViewById(R.id.btnCancelarActualizarUsuario);

        if (new PreferenciasHelper(getApplication()).obtenerToken() != null) {
            viewModel.obtenerMiUsuario(this, new UsuarioViewModel.UsuarioCallback() {
                @Override
                public void onSuccess(UsuarioDto usuario) {
                    editTextNombre.setText(usuario.getNombre());
                    editTextApellido.setText(usuario.getApellido());
                    editTextEmail.setText(usuario.getEmail());
                    editTextCelular.setText(usuario.getCelular());
                }

                @Override
                public void onError(String mensaje) {
                    Toast.makeText(UsuarioActivity.this, "Error al cargar usuario: " + mensaje, Toast.LENGTH_SHORT).show();
                }
            });
        }

        btnActualizarImagen.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                seleccionarImagen();
            }
        });

        btnActualizarDatos.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                String nombre = editTextNombre.getText().toString();
                String apellido = editTextApellido.getText().toString();
                String email = editTextEmail.getText().toString();
                String celular = editTextCelular.getText().toString();

                if (celular.length() == 9) {
                    UsuarioDto dtoUsuario = new UsuarioDto();
                    dtoUsuario.setNombre(nombre);
                    dtoUsuario.setApellido(apellido);
                    dtoUsuario.setEmail(email);
                    dtoUsuario.setCelular(celular);

                    if (imagenSeleccionadaUri != null) {
                        actualizarUsuarioConImagen(dtoUsuario, imagenSeleccionadaUri);
                    } else {
                        actualizarUsuario(dtoUsuario);
                    }

                    popupWindow.dismiss();
                } else {
                    Toast.makeText(UsuarioActivity.this, "El número de celular debe tener 9 dígitos.", Toast.LENGTH_SHORT).show();
                }
            }
        });

        btnCancelarActualizarUsuario.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                popupWindow.dismiss();
                startActivity(new Intent(UsuarioActivity.this, UsuarioActivity.class));
            }
        });

        popupWindow.showAtLocation(findViewById(R.id.activity_usuario), 17, 0, 0);
    }

    private void seleccionarImagen() {
        startActivityForResult(new Intent(Intent.ACTION_PICK, MediaStore.Images.Media.EXTERNAL_CONTENT_URI), 1);
    }

    @Override
    protected void onActivityResult(int requestCode, int resultCode, Intent data) {
        super.onActivityResult(requestCode, resultCode, data);
        if (resultCode == RESULT_OK && requestCode == 1) {
            imagenSeleccionadaUri = data.getData();
        }
    }

    private void actualizarUsuario(UsuarioDto dtoUsuario) {
        if (new PreferenciasHelper(getApplication()).obtenerToken() != null) {
            viewModel.actualizarMiPerfil(this, dtoUsuario, null, new UsuarioViewModel.ActualizarCallback() {
                @Override
                public void onSuccess(String mensaje) {
                    Toast.makeText(UsuarioActivity.this, mensaje, Toast.LENGTH_SHORT).show();
                    cargarUsuario();
                }

                @Override
                public void onError(String mensaje) {
                    Toast.makeText(UsuarioActivity.this, mensaje, Toast.LENGTH_SHORT).show();
                }
            });
        }
    }

    private void actualizarUsuarioConImagen(UsuarioDto dtoUsuario, Uri imagenUri) {
        if (new PreferenciasHelper(getApplication()).obtenerToken() != null) {
            viewModel.actualizarMiPerfil(this, dtoUsuario, imagenUri, new UsuarioViewModel.ActualizarCallback() {
                @Override
                public void onSuccess(String mensaje) {
                    Toast.makeText(UsuarioActivity.this, mensaje, Toast.LENGTH_SHORT).show();
                    cargarUsuario();
                }

                @Override
                public void onError(String mensaje) {
                    Toast.makeText(UsuarioActivity.this, mensaje, Toast.LENGTH_SHORT).show();
                }
            });
        }
    }

    private void cerrarSesion() {
        new PreferenciasHelper(getApplication()).limpiarPreferencias();
        Toast.makeText(this, "Sesión cerrada correctamente", Toast.LENGTH_SHORT).show();
        Intent intent = new Intent(this, MainActivity.class);
        intent.setFlags(Intent.FLAG_ACTIVITY_CLEAR_TOP);
        startActivity(intent);
        finish();
    }

    private void cargarUsuario() {
        if (new PreferenciasHelper(getApplication()).obtenerToken() != null) {
            viewModel.obtenerMiUsuario(this, new UsuarioViewModel.UsuarioCallback() {
                @Override
                public void onSuccess(UsuarioDto usuario) {
                    textNombre.setText(usuario.getNombre() + " " + usuario.getApellido());
                    textEmail.setText(usuario.getEmail());
                    textCelular.setText(usuario.getCelular());
                    Glide.with(UsuarioActivity.this).load(usuario.getUrlUsuario()).into(imageUsuario);
                }

                @Override
                public void onError(String mensaje) {
                    Toast.makeText(UsuarioActivity.this, "Error al cargar usuario: " + mensaje, Toast.LENGTH_SHORT).show();
                }
            });
        }
    }
}

File: app/src/main/java/com/diamond/appcliente/actividades/UsuarioActivity.kt
package com.diamond.appcliente.actividades

import android.content.Intent
import android.net.Uri
import android.os.Bundle
import android.provider.MediaStore
import android.util.Log
import android.view.Gravity
import android.view.View
import android.view.ViewGroup
import android.widget.Button
import android.widget.EditText
import android.widget.ImageView
import android.widget.PopupWindow
import android.widget.TextView
import android.widget.Toast
import androidx.activity.viewModels
import androidx.appcompat.app.AppCompatActivity
import androidx.lifecycle.Lifecycle
import androidx.lifecycle.lifecycleScope
import androidx.lifecycle.repeatOnLifecycle
import com.auth0.android.jwt.JWT
import com.bumptech.glide.Glide
import com.diamond.appcliente.R
import com.diamond.appcliente.dto.usuario.UsuarioDto
import com.diamond.appcliente.ui.state.UiState
import com.diamond.barbershop.shared.util.PreferenciasHelper
import com.diamond.appcliente.viewmodel.ListarReservaViewModel
import com.diamond.appcliente.viewmodel.UsuarioViewModel
import com.google.android.material.bottomnavigation.BottomNavigationView
import com.google.android.material.progressindicator.LinearProgressIndicator
import dagger.hilt.android.AndroidEntryPoint
import kotlinx.coroutines.launch
import javax.inject.Inject

@AndroidEntryPoint
class UsuarioActivity : AuthActivity() {

    private lateinit var btnLogoutUser: Button
    private lateinit var imageUsuario: ImageView
    private lateinit var textCelular: TextView
    private lateinit var textEmail: TextView
    private lateinit var textNombre: TextView
    private lateinit var textProgresoReservas: TextView
    private lateinit var progressRecompensa: LinearProgressIndicator
    private lateinit var textRecompensaDisponible: TextView
    private val viewModel: UsuarioViewModel by viewModels()
    private val listarReservaViewModel: ListarReservaViewModel by viewModels()
    private var imagenSeleccionadaUri: Uri? = null
    private var tvImagenSeleccionadaRef: TextView? = null
    private var layoutImagenSeleccionadaRef: android.view.View? = null
    private var btnSeleccionarImagenRef: android.widget.Button? = null

    @Inject lateinit var preferenciasHelper: PreferenciasHelper

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_usuario)

        imageUsuario = findViewById(R.id.imageUsuario)
        textNombre = findViewById(R.id.textNombre)
        textEmail = findViewById(R.id.textEmail)
        textCelular = findViewById(R.id.textCelular)
        btnLogoutUser = findViewById(R.id.btnLogoutUser)
        textProgresoReservas = findViewById(R.id.textProgresoReservas)
        progressRecompensa = findViewById(R.id.progressRecompensa)
        textRecompensaDisponible = findViewById(R.id.textRecompensaDisponible)

        val token = preferenciasHelper.obtenerToken()
        if (token != null) {
            Log.d("UsuarioActivity", "Token recibido: $token")
            JWT(token)
            viewModel.obtenerMiUsuario()
        } else {
            Log.d("UsuarioActivity", "Token no encontrado")
        }

        window.statusBarColor = -16777216

        findViewById<Button>(R.id.buttonActualizarUsuario).setOnClickListener { mostrarPopupActualizar() }
        btnLogoutUser.setOnClickListener { cerrarSesion() }

        val bottomNavigationView = findViewById<BottomNavigationView>(R.id.bottomNavigationView)
        bottomNavigationView.selectedItemId = R.id.nav_perfil
        bottomNavigationView.setOnItemSelectedListener { item ->
            when (item.itemId) {
                R.id.nav_home -> { startActivity(Intent(this, ClienteHomeActivity::class.java)); overridePendingTransition(0, 0); true }
                R.id.nav_servicios -> { startActivity(Intent(this, SeccionServiciosActivity::class.java)); overridePendingTransition(0, 0); true }
                R.id.historial -> { startActivity(Intent(this, ListarReservaActivity::class.java)); overridePendingTransition(0, 0); true }
                R.id.nav_perfil -> { startActivity(Intent(this, UsuarioActivity::class.java)); overridePendingTransition(0, 0); true }
                else -> false
            }
        }

        lifecycleScope.launch {
            repeatOnLifecycle(Lifecycle.State.STARTED) {
                launch {
                    viewModel.usuario.collect { state ->
                        when (state) {
                            is UiState.Success -> {
                                val usuario = state.data
                                textNombre.text = "${usuario.nombre} ${usuario.apellido}"
                                textEmail.text = usuario.email
                                textCelular.text = usuario.celular
                                Glide.with(this@UsuarioActivity).load(usuario.urlUsuario)
                                    .placeholder(R.drawable.perfil_default).into(imageUsuario)
                            }
                            is UiState.Error -> Toast.makeText(this@UsuarioActivity, state.message, Toast.LENGTH_SHORT).show()
                            else -> {}
                        }
                    }
                }
                launch {
                    viewModel.actualizarEvento.collect { msg ->
                        Toast.makeText(this@UsuarioActivity, msg, Toast.LENGTH_SHORT).show()
                    }
                }
                launch {
                    viewModel.error.collect { msg ->
                        Toast.makeText(this@UsuarioActivity, msg, Toast.LENGTH_SHORT).show()
                    }
                }
                launch {
                    listarReservaViewModel.reservas.collect { state ->
                        if (state is UiState.Success) {
                            actualizarProgresoRecompensa(state.data.count { it.estRecompensa == 0 })
                        }
                    }
                }
            }
        }
    }

    override fun onResume() {
        super.onResume()
        listarReservaViewModel.cargarReservas()
    }

    private fun actualizarProgresoRecompensa(reservasAcumuladas: Int) {
        val umbral = 7
        val count = reservasAcumuladas.coerceAtMost(umbral)
        val porcentaje = (count * 100) / umbral
        textProgresoReservas.text = "$count de $umbral reservas para tu próxima reserva gratuita"
        progressRecompensa.progress = porcentaje
        textRecompensaDisponible.visibility = if (reservasAcumuladas >= umbral) View.VISIBLE else View.GONE
    }

    private fun mostrarPopupActualizar() {
        val popupView = layoutInflater.inflate(R.layout.popup_actualizar_usuario, null)
        val popupWindow = PopupWindow(popupView, ViewGroup.LayoutParams.MATCH_PARENT, ViewGroup.LayoutParams.WRAP_CONTENT, true)

        val editTextNombre = popupView.findViewById<EditText>(R.id.etNombreUsuario)
        val editTextApellido = popupView.findViewById<EditText>(R.id.etApellidoUsuario)
        val editTextEmail = popupView.findViewById<EditText>(R.id.etEmailUsuario)
        val editTextCelular = popupView.findViewById<EditText>(R.id.etCelularUsuario)

        if (preferenciasHelper.obtenerToken() != null) {
            val u = (viewModel.usuario.value as? UiState.Success)?.data
            if (u != null) {
                editTextNombre.setText(u.nombre)
                editTextApellido.setText(u.apellido)
                editTextEmail.setText(u.email)
                editTextCelular.setText(u.celular)
            }
        }

        val btnSeleccionarImagen = popupView.findViewById<Button>(R.id.btnSeleccionarImagenUsuario)
        btnSeleccionarImagenRef = btnSeleccionarImagen
        tvImagenSeleccionadaRef = popupView.findViewById(R.id.tvImagenSeleccionada)
        layoutImagenSeleccionadaRef = popupView.findViewById(R.id.layoutImagenSeleccionada)
        btnSeleccionarImagen.setOnClickListener { seleccionarImagen() }

        popupView.findViewById<Button>(R.id.btnActualizarUsuario).setOnClickListener {
            val nombre = editTextNombre.text.toString()
            val apellido = editTextApellido.text.toString()
            val email = editTextEmail.text.toString()
            val celular = editTextCelular.text.toString()

            if (celular.length == 9) {
                val dtoUsuario = UsuarioDto().apply {
                    this.nombre = nombre
                    this.apellido = apellido
                    this.email = email
                    this.celular = celular
                }
                viewModel.actualizarMiPerfil(dtoUsuario, imagenSeleccionadaUri)
                popupWindow.dismiss()
            } else {
                Toast.makeText(this, "El número de celular debe tener 9 dígitos.", Toast.LENGTH_SHORT).show()
            }
        }

        popupView.findViewById<Button>(R.id.btnCancelarActualizarUsuario).setOnClickListener {
            popupWindow.dismiss()
            startActivity(Intent(this, UsuarioActivity::class.java))
        }

        popupWindow.setOnDismissListener {
            tvImagenSeleccionadaRef = null
            layoutImagenSeleccionadaRef = null
            btnSeleccionarImagenRef = null
        }
        popupWindow.showAtLocation(findViewById(R.id.activity_usuario), Gravity.CENTER, 0, 0)
    }

    private fun seleccionarImagen() {
        startActivityForResult(Intent(Intent.ACTION_PICK, MediaStore.Images.Media.EXTERNAL_CONTENT_URI), 1)
    }

    @Deprecated("Deprecated in Java")
    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        super.onActivityResult(requestCode, resultCode, data)
        if (resultCode == RESULT_OK && requestCode == 1) {
            val uri = data?.data
            imagenSeleccionadaUri = uri
            if (uri != null) {
                val fileName = uri.lastPathSegment?.substringAfterLast("/") ?: uri.toString().substringAfterLast("/")
                tvImagenSeleccionadaRef?.text = fileName
                layoutImagenSeleccionadaRef?.visibility = android.view.View.VISIBLE
                btnSeleccionarImagenRef?.text = "CAMBIAR IMAGEN"
            }
        }
    }

    private fun cerrarSesion() {
        getSharedPreferences("diamond_prefs", MODE_PRIVATE).edit().putBoolean("welcome_shown", false).apply()
        preferenciasHelper.limpiarPreferencias()
        Toast.makeText(this, "Sesión cerrada correctamente", Toast.LENGTH_SHORT).show()
        val intent = Intent(this, MainActivity::class.java)
        intent.flags = Intent.FLAG_ACTIVITY_CLEAR_TOP
        startActivity(intent)
        finish()
    }
}
